### ISSUE TITLE
PurgeLinesAndUnload not accounting for commented out purge moves in Start G-CODE

### ISSUE THREAD

**Issue by PartySausage** [GregValiant/GV-PostProcessors#2]
Hi Greg. I'm just trying out your latest iteration of the PurgeLinesAndUnload script & noticed that the script doesn't seem to be accounting for purge moves in the Start G-CODE that have already commented out & attempts to comment them again making the formatting look a bit odd in the GCODE. 
I think I mentioned this on a previous iteration of the script, which you sorted but the search syntax used was different then.

So I was wondering if something like this could be used

if "G1"in line and " E" in line and (" X" in line or " Y" in line) and not ( ";G1" in line or "; G1" in line) :

instead of

if "G1"in line and " E" in line and (" X" in line or " Y" in line):

You could no doubt make a neater job of it to account for a ';' at the start of a line

**Comment by GregValiant** [comment]
The formatting (I like things neat) makes the lines with double semi-colons get pushed to the right.
```
M109 S210                      ;Wait for extruder to reach print temp
G92 E0                         ;Reset Extruder
;G1 X67.5 Y0 Z0.4 F300         ;Move to start position
;                              ;G1 X167.5 E30 F400 ;Draw the first line
;                              ;G1 Z0.6 F120.0 ;Move to side a little
;                              ;G1 X162.5 F3000
G92 E0                         ;Reset Extruder
```

But comments are comments so functionally it doesn't matter.
I would probably opt for
`if "G1"in line and " E" in line and (" X" in line or " Y" in line) and not line.startswith(";") :`
That would exclude lines that are already comments.
I've noticed variations in the syntax in different printer definition files.  For Example, Elegoo may start with:
` G1 Fxxx Xxxx Yxxx`
The "filter" code was missing that because the line starts with a "space" character.  I'm concerned that if the filter is too specific it might comment out something that should be left in.  The script is commenting out everything from the first line that matches the criteria, to the first "G92 E0" line.
"PurgeLinesAndUnload" is in review for inclusion in Cura.  It will be a while though so I'll play with your suggested change and see if it isn't too exclusive.
(There is a typo in that line you quoted that I just fixed.  Did you notice it?

**Comment by GregValiant** [comment]
I updated the script using the "and not line.startswith(";")" form.
I'll update the pull request after I've played with it some more with different printer definitions.
I'll go ahead and close this as completed.  I might have to revert the code though depending on how it acts with other machines.

**Comment by PartySausage** [comment]
Hi Greg, Thanks for the script updates. Apologies for not getting back to you sooner; I’ve only had chance to play with the script this evening.
I thought there would be a command something like ‘startswith’ but couldn’t find a command in your script that seemed to fit the bill, then I did another search after getting your message & loads popped up…. like about two lines below the instruction doh!!
The difficult job for you is to try and account for a lot of different printers & startup configurations across thousands of users whereas I’m only dealing with one printer instance so I’m not expecting any sort of special treatment

In the start G-CODE for my E3 S1 PRO I have purge line commands configured for the Y (left) & X (front) axes & selected the one to use depending on whether I needed to use the extreme edges of the buildplate on that axis .
Prior to using you script I was using the one on the X axis & the Y axis commands commented out with a single semi-colon.
When I started using your script I commented out the X axis commands with a double semi colon ‘;;’ as below (For what reason?.... well there isn’t one really it was just a reminder to myself, which one I’d been using)

; Purge Line Y-Axis
;G1 X0.1 Y20.0 Z0.3 F5000.0 ; Move to start position

;G1 X0.1 Y200.0 Z0.3 F1500.0 E15 ; Draw the first line
;G1 X0.4 Y200.0 Z0.3 F5000.0 ; Move to side a little
;G1 X0.4 Y20.0 Z0.3 F1500.0 E30 ; Draw the second line

;G92 E0 ; Reset Extruder

;G1 Z2.0 F3000 ; Move Z Axis up little to prevent scratching of Heat Bed
;G1 X5 Y20 Z0.3 F5000.0 ; Move over to prevent blob squish

; Purge Line X-Axis
;;G1 X20.0 Y0.1 Z0.3 F5000.0 ; Move to start position

;;G1 X200.0 Y0.1 Z0.3 F1500.0 E20 ; Draw the first line
;;G1 X200.0 Y0.4 Z0.3 F5000.0 ; Move to side a little
;;G1 X20.0 Y0.4 Z0.3 F1500.0 E30 ; Draw the second line

;;G92 E0 ; Reset Extruder

;;G1 Z2.0 F3000 ; Move Z Axis up little to prevent scratching of Heat Bed
;;G1 Y5 X20 Z0.2 F5000.0 ; Move over to prevent blob squish

Using the original script the effect on the G-CODE was:

; Purge Line Y-Axis
;G1 X0.1 Y20.0 Z0.3 F5000.0          ; Move to start position
;                              ;G1 X0.1 Y200.0 Z0.3 F1500.0 E15         ; Draw the first line
;                              ;G1 X0.4 Y200.0 Z0.3 F5000.0             ; Move to side a little
;                              ;;G1 X0.4 Y20.0 Z0.3 F1500.0 E30         ; Draw the second line
;                              ;
;                              ;;G92 E0                     ; Reset Extruder
;                              ;
;                              ;;G1 Z2.0 F3000              ; Move Z Axis up little to prevent scratching of Heat Bed
;                              ;;G1 X5 Y20 Z0.3 F5000.0           ; Move over to prevent blob squish
;                              ;
;                              ;; Purge Line X-Axis
;                              ;;;G1 X20.0 Y0.1 Z0.3 F5000.0            ; Move to start position
;                              ;
;                              ;;;;G1 X200.0 Y0.1 Z0.3 F1500.0 E20             ; Draw the first line
;                              ;;;;G1 X200.0 Y0.4 Z0.3 F5000.0          ; Move to side a little
;                              ;;;;;G1 X20.0 Y0.4 Z0.3 F1500.0 E30             ; Draw the second line
;                              ;;;
;                              ;;;;;G92 E0                        ; Reset Extruder
;                              ;;;
;                              ;;;;;G1 Z2.0 F3000                 ; Move Z Axis up little to prevent scratching of Heat Bed
;                              ;;;;;G1 Y5 X20 Z0.2 F5000.0        ; Move over to prevent blob squish.

So I couldn’t understand why so many semi-colons were being appended to the lines, hence some additional digging.

With the revised script using the ‘startswith’ command the resulting G-CODE is:

; Purge Line Y-Axis
;G1 X0.1 Y20.0 Z0.3 F5000.0          ; Move to start position
;G1 X0.1 Y200.0 Z0.3 F1500.0 E15          ; Draw the first line
;G1 X0.4 Y200.0 Z0.3 F5000.0        ; Move to side a little
;G1 X0.4 Y20.0 Z0.3 F1500.0 E30           ; Draw the second line
;G92 E0                                          ; Reset Extruder
;G1 Z2.0 F3000                             ; Move Z Axis up little to prevent scratching of Heat Bed
;G1 X5 Y20 Z0.3 F5000.0                  ; Move over to prevent blob squish
; Purge Line X-Axis
;                              ;G1 X20.0 Y0.1 Z0.3 F5000.0        ; Move to start position
;                              ;G1 X200.0 Y0.1 Z0.3 F1500.0 E20         ; Draw the first line
;                              ;G1 X200.0 Y0.4 Z0.3 F5000.0             ; Move to side a little
;                              ;G1 X20.0 Y0.4 Z0.3 F1500.0 E30          ; Draw the second line
;                              ;G92 E0                      ; Reset Extruder
;                              ;G1 Z2.0 F3000               ; Move Z Axis up little to prevent scratching of Heat Bed
;                              ;G1 Y5 X20 Z0.2 F5000.0            ; Move over to prevent blob squish

So much improved, however with both the original script & the revised script with startswith instruction the formatting code interprets the 2nd semi-colon as the description text not the additional comment at the start of the command & indents the command rom the 2nd semi-colon onwards. The way I was thinking the formatting should work was that the only commands that should be indented should be those that have been modified by the script so as to differentiate them from the unmodified commands

My Python abilities are very..very limited & I think I’ve managed to sort the formatting code so that it doesn’t indent commands starting with a ; or ;; with the following addition

        if gap_len < 30: gap_len = 30
        for temp_index, temp_line in enumerate(temp_lines):
            if ";" in temp_line and not temp_line.startswith(";"):
                temp_lines[temp_index] = temp_line.replace(temp_line.split(";")[0], temp_line.split(";")[0] + str(" " * (gap_len - len(temp_line.split(";")[0]))),1)
            # This formats lines that are commented out but contain additional comments Ex:  ;M420 ; levelling mesh
            elif temp_line.startswith(";") and ";" in temp_line[1:] and not temp_line.startswith(";;") :
                temp_lines[temp_index] = temp_line[1:].replace(temp_line[1:].split(";")[0], ";" + temp_line[1:].split(";")[0] + str(" " * (gap_len - 1 - len(temp_line[1:].split(";")[0]))),1)
        any_gcode_str = "\n".join(temp_lines)

Which gives the following G-CODE’s for ;; & ; commented out lines but I’ve reached the limit of my abilities & can’t work out how to ident the commands that have been commented out by the script without affecting other commands in the startup

; Purge Line Y-Axis
;G1 X0.1 Y20.0 Z0.3 F5000.0          ; Move to start position
;G1 X0.1 Y200.0 Z0.3 F1500.0 E15          ; Draw the first line
;G1 X0.4 Y200.0 Z0.3 F5000.0        ; Move to side a little
;G1 X0.4 Y20.0 Z0.3 F1500.0 E30           ; Draw the second line
;G92 E0                                          ; Reset Extruder
;G1 Z2.0 F3000                             ; Move Z Axis up little to prevent scratching of Heat Bed
;G1 X5 Y20 Z0.3 F5000.0                  ; Move over to prevent blob squish
; Purge Line X-Axis
;;G1 X20.0 Y0.1 Z0.3 F5000.0        ; Move to start position
;;G1 X200.0 Y0.1 Z0.3 F1500.0 E20         ; Draw the first line
;;G1 X200.0 Y0.4 Z0.3 F5000.0             ; Move to side a little
;;G1 X20.0 Y0.4 Z0.3 F1500.0 E30          ; Draw the second line
;;G92 E0                      ; Reset Extruder
;;G1 Z2.0 F3000               ; Move Z Axis up little to prevent scratching of Heat Bed
;;G1 Y5 X20 Z0.2 F5000.0            ; Move over to prevent blob squish

&

; Purge Line Y-Axis
;G1 X0.1 Y20.0 Z0.3 F5000.0          ; Move to start position
;G1 X0.1 Y200.0 Z0.3 F1500.0 E15          ; Draw the first line
;G1 X0.4 Y200.0 Z0.3 F5000.0        ; Move to side a little
;G1 X0.4 Y20.0 Z0.3 F1500.0 E30           ; Draw the second line
;G92 E0                                          ; Reset Extruder
;G1 Z2.0 F3000                             ; Move Z Axis up little to prevent scratching of Heat Bed
;G1 X5 Y20 Z0.3 F5000.0                  ; Move over to prevent blob squish
; Purge Line X-Axis
;G1 X20.0 Y0.1 Z0.3 F5000.0          ; Move to start position
;G1 X200.0 Y0.1 Z0.3 F1500.0 E20          ; Draw the first line
;G1 X200.0 Y0.4 Z0.3 F5000.0        ; Move to side a little
;G1 X20.0 Y0.4 Z0.3 F1500.0 E30           ; Draw the second line
;G92 E0                                          ; Reset Extruder
;G1 Z2.0 F3000                             ; Move Z Axis up little to prevent scratching of Heat Bed
;G1 Y5 X20 Z0.2 F5000.0                  ; Move over to prevent blob squish

This above is a pretty niche case scenario & I’m not asking you to spend valuable time investigating a solution as I can just comment the commands out to save the script having to do it. It was just an explanation of how the script is functioning in my usage, which may be of use to you in the future.

What I did notice was that if I leave it up to the script to comment out the commands; The first G1 move command & the G1 commands after the G23 Extruder reset don’t get commented out, which from looking at the script I’m pretty sure that is intended as although they are related to the purge lines they are non extruding moves, which although unnecessary I can’t see a way for the script to comment them out without impacting any other G1 move commands in the startup. This in itself is a good enough reason for me to manually comment out the purge lines commands in my startup

; Purge Line Y-Axis
;G1 X0.1 Y20.0 Z0.3 F5000.0         ; Move to start position
;G1 X0.1 Y200.0 Z0.3 F1500.0 E15          ; Draw the first line
;G1 X0.4 Y200.0 Z0.3 F5000.0        ; Move to side a little
;G1 X0.4 Y20.0 Z0.3 F1500.0 E30           ; Draw the second line
;G92 E0                                         ; Reset Extruder
;G1 Z2.0 F3000                            ; Move Z Axis up little to prevent scratching of Heat Bed
;G1 X5 Y20 Z0.3 F5000.0                 ; Move over to prevent blob squish
; Purge Line X-Axis
G1 X20.0 Y0.1 Z0.3 F5000.0           ; Move to start position
;G1 X200.0 Y0.1 Z0.3 F1500.0 E20          ; Draw the first line
;G1 X200.0 Y0.4 Z0.3 F5000.0        ; Move to side a little
;G1 X20.0 Y0.4 Z0.3 F1500.0 E30           ; Draw the second line

G92 E0                                           ; Reset Extruder
G1 Z2.0 F3000                              ; Move Z Axis up little to prevent scratching of Heat Bed
G1 Y5 X20 Z0.2 F5000.0             ; Move over to prevent blob squish

Apologies for the rather long winded message. I hope you managed to stay awake until the end. If you need me to test anything in the future drop me a message

Oh almost forgot…. As for the typo, if it’s the space missing from “G1”in line then yes I did spot it & corrected it but forgot to mention in my message…. Another Doh!

Regards Graeme

**Comment by GregValiant** [comment]
If you are going to use the script then you can just eliminate the lines from the StartUp Gcode rather than comment them out.

In the script - You can also put a # sign in front of lines 194 and 195 to disable the formatting.
        # data[1] = self._format_string(data[1])   # This one formats the start up
        # data[-1] = self._format_string(data[-1])  # this one is for the ending Gcode
I just threw those in to make things more readable.  I look at so many gcodes from the Cura Github site I figured mine could at least look neat and readable.

Out of all of this crap the three scripts I keep active for every print are Purge Lines, Advanced Fan Control, and Display Info on LCD.  The rest come and go depending on the model, but the three prime ones stay.

Like I mentioned, I don't want to "over filter" because something is sure to get missed.  Because all the StartUp Gcodes are different, it's a dance.

FOr instance, this is my StartUp for my Ender 3 Pro:

```
;--------------------------StartUp Gcode
M113 S0                    ; No Echo's
M203 E{machine_max_feedrate_e}                       ; Set max E speed
{if 'TPU' in material_type}
M92 E119                   ; Steps/mm for TPU
{elif 'PETG' in material_type}
M92 E102                   ; Steps/mm for PETG
{else}
M92 E97                    ; Steps/mm for PLA
{endif}
{if machine_center_is_zero}
M206 X-116 Y-121 Z0     ; Home offsets 'Origin at center'
{else}
M206 X-1 Y-6 Z0         ; Home offsets Origin LF corner
{endif}
G21                        ; Metric Units
G90                        ; Absolute Movement
M82                        ; Absolute Extrusion
M220 S100                  ; Reset FeedRate
M221 S100                  ; Reset FlowRate
M140 S{material_bed_temperature_layer_0}                   ; Start heating the bed
{if material_bed_temperature_layer_0 > 65}
M104 S200                  ; Standby while bed heats
M190 S{material_bed_temperature_layer_0}                   ; Bed wait if over 65
{endif}
M109 S{material_print_temperature_layer_0}                 ; Hot End Temperature
G92 E0                     ; Reset E
G28 X Y                    ; Home XY axes
G0 F600 Z2                 ; Move up
;--------------------------End of StartUp
```
Trying to filter that mess would be tough.  A lot of people have custom Start gcodes.  The code can only account for so much.
(The G28 X Y is because I level with a piece of paper and I don't want the Z to home once I finish leveling.)

**Comment by PartySausage** [comment]
Hi Greg, I will probably just remove the commands from my startup, having got the confidence over numerous prints that the script is working well for my setup.

I have several of your scripts configured on my installation & just enable/disable them depending on my requirements for the current print but this one has been in constant use. So thanks for developing some excellent utilities

The one thing I have noticed is that on my setup the Purge Lines script and the SpoonTabs script didn’t play well together if the Purge Lines script was higher than the SpoonTabs script in the Post-Processor list.
The result was that nozzle was too high above the bed when it came to print, which is understandable with both scripts modifying G-code output & the SpoonTabs script changing how the first layer would be printed to print the tabs first after the Pause Lines script had done it’s G-code mods so I just make sure that the Purge Lines script is the last script in the list of any scripts that modify the print portion of the G-code & all works well 😊

Like a dog with a bone I’ve been using this script to expand my knowledge of Python & I’ve sort of produced something that formats the start/end G-codes, whilst identifying, which commands have been disabled by the Purge Lines script
I did this by appending a ‘#’ to the ‘;’ which is added when a command to be disabled has been identified. The ‘#’ is then removed & and a message is added in it’s place when the line is formatted.
I also added some code to preserve the blank lines that are in my start/end G-codes between some of the commands so it looks similar in the G-Code file.
The final bit was to add some additional formatting strings to account for the additional combinations

It’s probably a bit verbose & not very efficiently coded but I’ve attached the script if you were interested in having a looksie (laugh).

All the above resulted in the following in my G-code:

; *** Ender 3 S1 Pro Start G-code ***

M413 S0                                       ; Disable power loss recovery

G92 E0                                        ; Reset Extruder

; *** Preheat Build-Plate & Hotend ***

M140 S60                                      ; Start heating Build-Plate
M109 S215.0                                   ; Start heating Hotend & wait for Hotend to reach temperature
M190 S60                                      ; Wait for Build-Plate to reach temperature

G28                                           ; Home all axes

M420 S1 Z10                                   ; Enable ABL Mesh, Fade 10 Layers

M503                                          ; Send Current Printer Settings to Serial Monitor
M420 V1                                       ; Send ABL Mesh to Serial Monitor

;G1 Z10.0 F3000                               ; Move Z Axis up little to prevent scratching of Heat Bed

; *** Purge Line Y-Axis ***

;G1 X0.1 Y20.0 Z0.3 F5000.0                   ; Move to start position

;G1 X0.1 Y200.0 Z0.3 F1500.0 E15              ; Draw the first line
;G1 X0.4 Y200.0 Z0.3 F5000.0                  ; Move to side a little
;G1 X0.4 Y20.0 Z0.3 F1500.0 E30               ; Draw the second line

;G92 E0                                       ; Reset Extruder

;G1 Z2.0 F3000                                ; Move Z Axis up little to prevent scratching of Heat Bed
;G1 X5 Y20 Z0.3 F5000.0                       ; Move over to prevent blob squish

; *** Purge Line X-Axis ***

G1 X20.0 Y0.1 Z0.3 F5000.0                    ; Move to start position

;     >>> Command Disabled by the 'Purge Lines and Unload Filament' Post-Processor <<<     G1 X200.0 Y0.1 Z0.3 F1500.0 E20              ; Draw the first line
;     >>> Command Disabled by the 'Purge Lines and Unload Filament' Post-Processor <<<     G1 X200.0 Y0.4 Z0.3 F5000.0                  ; Move to side a little
;     >>> Command Disabled by the 'Purge Lines and Unload Filament' Post-Processor <<<     G1 X20.0 Y0.4 Z0.3 F1500.0 E30               ; Draw the second line

G92 E0                                        ; Reset Extruder

G1 Z2.0 F3000                                 ; Move Z Axis up little to prevent scratching of Heat Bed
G1 Y5 X20 Z0.2 F5000.0                        ; Move over to prevent blob squish
G92 E0
;TYPE:CUSTOM----------[Purge Lines at MinY]
G0 F600 Z2                                    ; Move up
G92 E0                                        ; Reset extruder
G0 F9000.0 X11 Y1                             ; Move to start
G0 F600 Z0.3                                  ; Move down
G1 F2700 X209 Y1 E12.47256                    ; First line
G0 X209 Y4                                    ; Move over
G1 F2700 X11 Y4 E24.94512                     ; Second line
G1 F2400 E24.14512                            ; Retract
G0 F600 Z8                                    ; Move Up
G4 S1                                         ; Wait for 1 second
G0 F2700 X21 Y4 Z0.3                          ; Slide over and down
G0 X36 Y4                                     ; Wipe
G0 F600 Z2                                    ; Move Z
;---------------------[End of Purge]
G92 E0
G1 F2400 E-0.8

;MESH:NONMESH---------[Travel to Layer Start]
G0 F600 Z2                                    ; Move up
G0 F9000.0 X6                                 ; Move
G0 F600 Z0                                    ; Nail down the string
G0 Z2                                         ; Move up
G0 F9000.0 Y6 Z2                              ; Move
G0 F600 Z0                                    ; Nail down the string
G0 Z2                                         ; Move up
G0 F9000.0 X214 Z2                            ; Ortho move
G0 F600 Z0                                    ; Nail down the string
G0 Z2                                         ; Move up
G0 F9000.0 Y133.291 Z2                        ; Ortho move
G0 F600 Z0                                    ; Nail down the string
G0 Z2                                         ; Move up
;---------------------[End of layer start travels]

**Comment by GregValiant** [comment]
I'm not a fan of the interfaces of the Slic3r clones like PrusaSlicer and Orca.  Cura's controls (dialog boxes, buttons, checkboxes, et al) are clumsy (_stupid_ might be a better word), but the settings are all together.
I've tried writing a couple of post processors for Prusa, but the way the code goes back and forth to the slicer seems to make it very limited in what can actually be done.  (Or maybe I'm just no knowledgeable enough).
Cura is really customizable and I like that a lot.

"Little Utilities" with it's numerous scripts, was difficult to put in order.  The user can't just tell it which order the scripts should run in because the order is hard coded.
When "Advanced Fan Control" is added to the Post Processing list, it actually checks to see if there is another instance of itself already loaded.  It uses that info to disable the "Remove M106 lines" checkbox if it isn't the first instance.  Something like that is of limited use because Cura doesn't always update setting changes and because a post processor doesn't run until everything else is done, and sending messages to the user letting them know when there is a conflict isn't a smooth process.

Anyway, I think the changes in the handling of the StartUp and Ending codes (that allow some logic) is really good.  There are a couple of bugs in it yet, but they are minor and will get worked out.

_Hellaholic_ is reviewing "PurgeLinesAndUnload" and no doubt will be suggesting changes.  I think the reason he cherry-picked it out of all the Pull Requests was that it is something he thought Cura should have as well.  For myself, I just want any print to start in an orderly fashion.  I think I've got it down pat and it only took me 4 years!

Have a good holiday.

**Comment by PartySausage** [comment]
I’ve never tried Orca but do have PrusaSlicer. I don’t use it for printing but it has some very good basic model manipulation & STL repair utilities that are lacking in Cura, which makes it easier to isolate parts of a model to do test prints without importing or having to do an edit in the CAD S/W. I agree re. the UI though, hence using Cura for the meat of the slicing and printing work

I have downloaded Little Utilities but haven’t had any need to use any of the functions yet but can see how it could be an implementation headache.

I haven’t investigated using the new logic in my start & end g-codes yet as again I haven’t felt that I’ve needed too. I’m sure it would help with some of my workflow but my printer seems to be working quite well at the moment so don’t want to break anything and apart from a couple of cura UI niggles that seems to be working reasonably well too.

I noticed that Hellaholic is reviewing your latest pull request & it will be interesting to see what mods he suggests. TBH despite my tweaks, I’m still quite happy with the way you've got it working currently 🙂

Enjoy the festivities 🎄
